src/workloads/scale/MemoryStress.yml

```diff
@@ -0,0 +1,90 @@
+SchemaVersion: 2018-07-01
+Owner: Service Architecture
+Description: |
+  Used to test performance of the server under heavy memory pressure.
+
+GlobalDefaults:
+  dbname: &DBName memorystress
+  MaxPhases: &MaxPhases 20
+
+  LoadThreads: &LoadThreads 128
+  LoadBatchSize: &LoadBatchSize 500
+
+  NumberOfDocuments: &NumDocs 20096
+  # NumberOfDocumentsPerWorker = NumberOfDocuments / LoadThreads.
+  NumberOfDocumentsPerWorker: &DocsPerThread 157
+  DocSize: &DocSize 520000
+  Document: &Doc
+    # Document number/ID.
+    a: {^Inc: {start: -157, step: 1, multiplier: *DocsPerThread}}
+    # Random number from [0, NumberOfDocuments].
+    b: {^RandomInt: {min: 0, max: *NumDocs}}
+    # Fill 520 KB.
+    c: {^RandomString: {length: *DocSize}}
+
+  # Number of documents to sort by each thread. Tuned so that *SortStep * *DocSize < 100 MB,
+  # which is the maximum memory a sort query is allowed to use before spilling to disk.
+  SortStep: &SortStep 180
+  SortBatchSize: &SortBatchSize 180
+  SortThreads: &SortThreads 250
+  SortRepeat: &SortRepeat 5
+
+Clients:
+  Default:
+    QueryOptions:
+      socketTimeoutMS: -1
+      maxPoolSize: 500
+    
+Actors:
+- Name: LoadDocuments
+  Type: Loader
+  Threads: *LoadThreads
+  Phases:
+    OnlyActiveInPhases:
+      Active: [0]
+      NopInPhasesUpTo: *MaxPhases
+      PhaseConfig:
+        CollectionCount: 1
+        Database: *DBName
+        Repeat: 1
+        Document: *Doc
+        MultipleThreadsPerCollection: true
+        DocumentCount: *NumDocs
+        BatchSize: *LoadBatchSize
+
+- Name: SortMany
+  Type: RunCommand
+  Threads: *SortThreads
+  Phases:
+    OnlyActiveInPhases:
+      Active: [1]
+      NopInPhasesUpTo: *MaxPhases
+      PhaseConfig:
+        Repeat: *SortRepeat
+        Database: *DBName
+        Operations:
+        - OperationMetricsName: SortMany
+          OperationName: RunCommand
+          OperationCommand:
+            # Loader default collection name.
+            aggregate: Collection0
+            # Sort *SortStep number of documents starting from a randomly chosen 'a'.
+            pipeline:
+              [{$match:
+                {$expr:
+                  {$let:
+                    {vars:
+                      {start: {^RandomInt: {min: 0, max: *NumDocs}}},
+                    in: {$and: [$gte: ["$a", "$$start"], $lt: ["$a", {$add: ["$$start", *SortStep]}]]}
+                    }
+                  }
+                }
+              },
+              {$sort: {b: 1}}]
+            cursor: {batchSize: *SortBatchSize}
+
+# AutoRun:
```

Comment:
As an effect of the purpose described in `Description` in the header, the memory heavy operations should cause the hosts associated with the latest mongod binaries to crash. This is because we currently do not have anything to prevent memory overload.

### Why commented out

The ideal purpose for putting this crash inducing workload into master is so that other server engineers can find this as an Evergreen task and run to test their implementations against. But because this would cause red/purple-ness on sys-perf with the current mongod binaries, I want to leave this workload commented out. Future engineers who want to use this workload would simply uncomment and run `schedule_patch_auto_tasks`.

In the future, I expect this to be uncommented as mongod gets better at handling memory pressure (and therefore would no longer experience these failures). Let me know what you think about this approach of commenting the workload out.

### Alternatives

Other options:
(1) Use task annotations to mark the workload as non problematic.
(2) Leave as YML file linked to the ticket, and engineers wishing to run the workload would put the file into `src/workloads/scale/` and run `schedule_patch_auto_tasks`, no merging to master required.
(3) Create a new temporary (or permanent) project with its own configurations (e.g. copy something like `system_perf.yml` and configure post run actions, run as `evergreen patch -p <new_project_name>`). This may be a bit more heavy handed but it would not cause any noise on `sys-perf` master.

Please do let me know if there are better ways of doing any of this.

Reply:
Will look into this, thanks!

Reply:
I've created EVG-21102 to look into exception handling ... but I suspect it is a relatively large feature so I wouldn't depend or wait on it.

Reply:
> From [here](https://github.com/evergreen-ci/evergreen/blob/ff46048d63feb595caedb68fd4d9fc85ab73a96b/model/project_parser.go#L435) and [here](https://github.com/evergreen-ci/commit-queue-sandbox/blob/b92cc103abfbef75b1529fde02952b0b48570e99/evergreen.yml#L42C9-L42C24) I think we can add a task to a variant so that it can only be executed manually:

Added a `memory_overload` task (naming must be different to avoid conflict with schedule_global_auto_tasks AKA generate.tasks, which will create MemoryStress.yml -> memory_stress), PR: https://github.com/10gen/mongo/pull/16354

Note: I believe the conventional way to select workloads to run is to use the [AutoRun feature](https://jira.mongodb.org/browse/TIG-2041) with the intention of making workloads easy to run without changing or adding a bunch of project configurations. Perhaps we could keep this task temporarily while it doubles as a performance and correctness test? Then as we add features and testing to prevent OOM, we can remove this as a special task (instead just running with the `schedule_patch_auto_task` like the other workloads)?

Reply:
> Note: I believe the conventional way to select workloads to run is to use the [AutoRun feature](https://jira.mongodb.org/browse/TIG-2041) with the intention of making workloads easy to run without changing or adding a bunch of project configurations. Perhaps we could keep this task temporarily while it doubles as a performance and correctness test? Then as we add features and testing to prevent OOM, we can remove this as a special task (instead just running with the `schedule_patch_auto_task` like the other workloads)?

Generally it is preferable to use an AutoRun ... but it is perfectly valid to create an explicit task and run it with more control (to control how and when it runs).

So your current approach is perfectly valid.

I had assumed you might want to set `activate: false` on the sys perf task ... but you seem to be suggesting that you want to actively develop and run it. **Is that correct?**

If this is the case then ... you should let the buildbarons know (on the \#perf-build-baroning slack channel) that  this test is expected to fail with a system failure. It looks like there can be 2 different system failure types so we should let them know about that also.

It would also make sense to see if we can get EVG-21102 started as:

1.  this would remove the system exception
2. should allow us to add the exception count (or whatever metrics is created) to signal processing long term.

Reply:
also I only just now noticed your question on evergreen-users about the name. Yes it is the way that the genny file name gets translated into a name that then clashes with the test name. I generally try to use '-' rather than underscores in system_perf.yml.

So:
```
  - name: memory-stress
    commands:
      - func: f_run_dsi_workload
        vars:
          test_control: auto_genny_workload
          auto_workload_path: ./src/genny/dist/etc/genny/workloads/scale/MemoryStress.yml
```

That is probably preferrable because otherwise it can be confusing when you are looking for the file and or test.

Reply:
I think this is intended only for local development, and shouldn't be pushed into the mainline waterfall (since we don't want to purposely push tests that will show up as failures, even if it's meant to happen).

Reply:
So with the project configuration here: https://github.com/10gen/mongo/pull/16354, any developer can run this as a checkable box task. The cron will also not periodically run this task because of `activate: false`, and therefore there shouldn't really be any alarm. I will also check in with #perf-build-baron to notify that this task is expected to fail in case someone accidentally runs this task. IMO, this doesn't really cause any noise and it just makes it easier for workload management people to test out any solutions/POCs. Is this an acceptable solution?

Otherwise we could just do the commenting out AutoRun, which IMO would be a bit more confusing for developers who don't use genny. lmk your thoughts

Reply:
I think we probably would rather avoid having the box checkable by anyone if it's going to fail just to avoid any confusion. I'd rather the workflow be that for the small amount of people who know that they want this particular workload (and so know what the intended task status is) need to do the manual step of checking out genny and uncommenting some lines in code.

I want to avoid the situation where people may decide to blindly run some set of workloads and then spend time digging into why a task is failing.

Reply:
Okay that makes sense. I will add a comment in the JIRA ticket after this is merged to basically to this: uncomment the lines, run evg patch sys-perf, then check `schedule_patch_auto_task` on the shard-lite variant.

Reply:
what ever works for you guys.